api/src/pcapi/routes/backoffice_v3/forms/custom_reimbursement_rule.py

```diff
@@ -31,3 +38,53 @@ def is_empty(self) -> bool:
                 self.offerer.data,
             )
         )
+
+
+class EditCustomReimbursementRule(FlaskForm):
+    offerer = fields.PCTomSelectField(
+        "Structure",
+        multiple=False,
+        choices=[],
+        validate_choice=False,
+        endpoint="backoffice_v3_web.autocomplete_offerers",
+    )
+    subcategories = fields.PCSelectMultipleField(
+        "Sous-catégories", choices=map(lambda s: (s.id, s.pro_label), subcategories_v2.ALL_SUBCATEGORIES)
+    )
+    rate = fields.PCStringField("Taux de remboursement %")
+    start_date = fields.PCDateField("Date de début d'application")
+    end_date = fields.PCDateField(
+        "Date de fin d'application (optionnelle)", validators=(wtforms.validators.Optional(),)
+    )
+
+    def validate_start_date(self, start_date: fields.PCDateField) -> fields.PCDateField:
+        end_date = self._fields["end_date"]
+
+        if end_date.data and (start_date.data > end_date.data):
+            raise wtforms.ValidationError("Ne peut pas être postérieure à la date de fin")
+
+        if start_date.data < (date.today() + timedelta(days=1)):
+            raise wtforms.ValidationError("Ne peut pas commencer avant demain")
+
+        start_date.data = datetime.combine(start_date.data, datetime.min.time())
+        return start_date
+
+    def validate_end_date(self, end_date: fields.PCDateField) -> fields.PCDateField:
+        if end_date.data:
+            end_date.data = datetime.combine(end_date.data, datetime.max.time())
+        return end_date
+
+    def validate_rate(self, rate: fields.PCDateField) -> fields.PCDateField:
+        str_value = rate.data.replace(",", ".")
+
+        if not re.match(r"^\d{1,2}.?\d{1,2}$", str_value):
```

Comment:
Attention, le `.` sans échappement correspond à tout caractère dans l'expression régulière, donc `12a34` va passer, et une exception se produira dans la conversion en `float`. Donc au minimum :
```suggestion
        if not re.match(r"^\d{1,2}\.?\d{1,2}$", str_value):
```
Pour rendre la partie décimale optionnelle sans accepter `1234` (puisque le point au milieu est optionnel), je suggère aussi cette forme :
```suggestion
        if not re.match(r"^(?:100|\d{1,2}(?:\.\d{1,2})?)$", str_value):
```
qui a priori doit empêcher les valeurs `> 100` ou `< 0`.

Reply:
Juste une autre idée, as-tu essayé simplement en donnant à l'attribut rate : 
- `coerce=float`
- le validateur `NumberRange` issu de wtforms ?

Reply:
non  parcequ'on veut accepter les virgules (sinon j'aurais juste mis un champ float sans m'embetter)

Reply:
Tu dois pouvoir remplacer la virgule par le point dans un filtre, passé dans `filters=...` au `StringField`, et exécuté avant la validation (je ne suis plus 100% sûr donc à vérifier)

Reply:
Je rejoins Patrick dans l'idée d'essayer de ne pas utiliser de regexp. Pourquoi ne pas faire simplement ce qui est actuellement fait dans Flask-Admin : 

https://github.com/pass-culture/pass-culture-main/blob/cd40c29cdfd3e90529aa264eb68a6b4bd9e7a540/api/src/pcapi/admin/custom_views/custom_reimbursement_rule_view.py#L286

On utilise `form.rate.data.replace(",", ".")`, on ajoute un try/except autour et on serait ok, non ?

Reply:
La regexp est la pour le cas NaN. tout les autres cas sont gérable bine plus facilement (et sans regexp).

Le probleme c'est e gérer la virgule française qui oblige a mettre un stringfield au lieu d'un type float.

Je ne peux pas non plus utiliser de filters a moin de redefinir completement ce qu'est un stringField (il refuse l'argument en disant qu'on lui a deja passé)

Reply:
Ma version lève une erreur si l'utilisateur transmet "NaN", puisqu'on tente `form.rate.data / 100`.

Reply:
j'ai pris une troisieme voies, simplement cast en float et vérifier manuellement le cas de NaN

Reply:
Je vois que dans FA on utilise un `DecimalField` avec `use_locale=True`, n'est-ce pas cela qui gère la virgule comme séparateur, justement parce qu'on travaille en français ?
Ça vaut peut-être la peine de suivre cette piste, en créant un : 
```
class PCDecimalField(wtforms.DecimalField):
    widget = partial(widget, template="components/forms/string_field.html")
    use_locale=True  # vérifier si ça marche ici et pas seulement dans l'__init__
```

PS : faut pas trop s'emballer sur le cas `NaN`, c'était pour être pointilleux, mais si les équipes métiers saisissent `NaN`... tant pis !